src/widgets/modals/ByProgramModal.js

```diff
@@ -1,376 +1,248 @@
-          currentStepKey: 'complete',
-        };
-        break;
-      default:
-        break;
-    }
-
-    this.setState(nextState, () => this.setCurrentSteps());
-  };
-
-  getSearchModalProps = () => {
-    const { programs, suppliers, orderTypes, periods, program, orderType } = this.state;
-
-    const getBaseProps = key => ({
-      queryString: 'code BEGINSWITH[c] $0',
-      sortByString: 'name',
-      primaryFilterProperty: 'name',
-      renderLeftText: item => `${item.name}`,
-      onSelect: selectedItem => this.setNewValue({ key, selectedItem }),
-    });
-
-    return {
-      program: {
-        ...getBaseProps('program'),
-        options: programs,
-        queryStringSecondary: 'name BEGINSWITH[c] $0',
-        renderRightText: item =>
-          item.parsedProgramSettings ? item.parsedProgramSettings.elmisCode : '',
-      },
-      supplier: {
-        ...getBaseProps('supplier'),
-        options: suppliers,
-        secondaryFilterProperty: 'code',
-        renderRightText: item => `(${item.code})`,
-      },
-      orderType: {
-        ...getBaseProps('orderType'),
-        options: orderTypes,
-        renderRightText: item =>
-          `${this.LOCALIZATION.misc.maxMOS}: ${item.maxMOS} - ${
-            this.LOCALIZATION.misc.thresholdMOS
-          }: ${item.thresholdMOS} - ${this.LOCALIZATION.misc.maxOrdersPerPeriod}: ${
-            item.maxOrdersPerPeriod
-          }`,
-      },
-      period: {
-        ...getBaseProps('period'),
-        options: periods,
-        renderRightText: item =>
-          `${item.toString()} - ${item.requisitionsForOrderType(program, orderType)}/${
-            orderType.maxOrdersPerPeriod
-          } ${this.LOCALIZATION.misc.requisitions}`,
-      },
-    };
-  };
-
-  renderTextEditor = () => {
-    const { name } = this.state;
+  // Modal component, for user selection
+  const ByProgramSelector = () => {
+    const { isModalOpen, currentKey } = state;
     return (
-      <TextEditor
-        text={name}
-        onEndEditing={selectedItem => this.setNewValue({ key: 'name', selectedItem })}
-      />
+      <PageContentModal isOpen={isModalOpen} onClose={onCloseModal} coverScreen>
+        {currentKey !== 'name' ? (
+          <AutocompleteSelector
+            queryString="code CONTAINS[c] $0"
+            sortByString="name"
+            primaryFilterProperty="name"
+            renderLeftText={item => `${item.name}`}
+            options={modalData}
+            {...modalProps({ dispatch, program, orderType })[currentKey]}
+          />
+        ) : (
+          <TextEditor text={name} onEndEditing={value => dispatch(setName(value))} />
+        )}
+      </PageContentModal>
     );
   };
 
-  renderModal = () => {
-    const { currentStepKey } = this.state;
-    if (currentStepKey === 'name') return this.renderTextEditor();
-    return <AutocompleteSelector {...this.getSearchModalProps()[currentStepKey]} />;
+  const stepProps = {
+    program: useMemo(
```

Comment:
Could potentially add a helper with defaults:

```
const getStepProps = stepProps => ({
  onPress: onOpenModal,
  type: 'select',
  field: 'name',
  ...stepProps,
});

const stepProps = {
  program: useMemo(
    getStepProps({
      data: program,
      getModalData: getPrograms,
      status: getStatus('program'),
      stepKey: 'program',
    }),
    [program, state.isProgramBased]
  ),
  ...
};
```

Reply:
how it kinda was before refactor. I think I prefer it but it offers little and is harder to read the benefits of being dry aren't so big when the code is in the same place anyway